django_pg_simple_hll/migrations/0001_initial.sql

```diff
@@ -0,0 +1,145 @@
+-- The state transition function
+-- it takes the current state, and the two input values
+-- hll_agg_state is the current running state, made up of bucket keys and hashes
+--  it is an array of n_buckets + 2
+-- input is the element we are considering
+-- hll_precision is the precision we are using for the approximation
+--  it use used to calculate the number of buckets
+create or replace function
+hll_bucket(hll_agg_state int[], input anyelement, hll_precision int)
+returns int[]
+language plpgsql immutable strict parallel safe as
+$$
+declare
+    n_buckets int := POW(2, hll_precision); -- eg.. for 512 buckets, hll_precision is 9 because 2**9 -> 512
+    hashed_input int := hashtext(input::text); -- transform the input into a text and hash it
+    bucket_hash int := hashed_input & ~(1 << 31); -- keep all the significant bits, without the sign
```

Comment:
I'm possibly reading this wrong.  But does this mean if `hashed_input` looks like `(b1)(b2) (b3)...(bn)` were each b1..bn is a bit, then `bucket_hash` will look like `0(b2)(b3)...(b31)00...00`?

Reply:
or actually perhaps I mean `bucket_hash` will look like `(b n-30)(b n-29)...b(n)`.  You just keep the last 31 bits?

Reply:
this keeps the rightmost 31 bits, and sets the "sign" bit to 1
`1(b1)(b2)...(b32)`

In python 
```python
f"{-1 & ~(1 << 31) & 0xffffffff:0b}" -> '1111111111111111111111111111111'
f"{(2**32 -1)  & ~(1 << 31) & 0xffffffff:0b}" -> '1111111111111111111111111111111'
```

(python int's have variable length that's why you have to do the `& 0xffffffff` bit